### PR TITLE
proxy: fix forward auth, request signing

### LINKED ISSUE #363: proxy: no redirection after successful authn/authz in forward-auth mode
## What happened?
I am using Traefik forward-auth middleware: `traefik.frontend.auth.forward.address=https://fwdauth.example.com/.pomerium/verify/whoami.example.com`, there is no redirection to orginal URL after successful authn/authz. ( just `foo@test.com is authorized for whoami.example.com` message in plaintext is shown. URL: whoami.example.com is accessible )

## What did you expect to happen?

**Flow:**
1. Hit URL: whoami.example.com
2. Traefik -> forward-auth redirection
3. Pomerium successful auth/authz
4. Redirect back to whoami.example.com

## What's your environment like?

- Pomerium version (retrieve with `pomerium --version` or `/ping` endpoint):
pomerium/v0.4.0 (+github.com/pomerium/pomerium; 28eae36; go1.13.1)
- Server Operating System/Architecture/Cloud:
Ubuntu 19.04 / Docker CE 19.03.1
## What's your config.yaml?
```
SERVICES="all" 
AUTHENTICATE_SERVICE_URL=https://auth.example.com/ 
FORWARD_AUTH_URL=https://fwdauth.example.com/ 
INSECURE_SERVER=TRUE 
POMERIUM_DEBUG=true  
COOKIE_SECRET=xxx 
IDP_CLIENT_ID=xxxx 
IDP_CLIENT_SECRET=xxxx 
IDP_PROVIDER=oidc 
IDP_PROVIDER_URL=https://xxxx.eu.auth0.com/ 
```
```config.yaml
policy:
  - from: https://whoami.example.com
    to: https://whoami.example.com
    allowed_users:
      - foo@example.com
```

## What did you see in the logs?

```logs
7:53PM DBG http-request duration=0.73882 email=foo@example.com fwd_ip=["89.176.XX.XX"," 172.18.0.1"] group= host=fwdauth.example.com ip=172.18.0.27 method=GET path=/.pomerium/verify/whoami.example.com req_id=cb7b36d6-b1f4-c246-070c-1b6bcfc91731 service=all size=53 status=200 user_agent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/77.0.3865.90 Safari/537.36"
7:53PM DBG http-request duration=0.898042 email=foo@example.com fwd_ip=["89.176.XX.XX"," 172.18.0.1"] group= host=fwdauth.example.com ip=172.18.0.27 method=GET path=/.pomerium/verify/whoami.example.com referer=https://whoami.example.com/ req_id=09cd803e-59f7-7d12-ad18-617fc139caed service=all size=53 status=200 user_agent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/77.0.3865.90 Safari/537.36"
7:53PM DBG proxy: re-authenticating due to session state error
7:53PM DBG http-request duration=0.156366 email= fwd_ip=["89.176.XX.XX"," 172.18.0.1"] group= host=fwdauth.example.com ip=172.18.0.27 method=GET path=/.pomerium/verify/whoami.example.com req_id=1f5e84dc-e654-0f4e-84fa-12a370d9d261 service=all size=233 status=302 user_agent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/77.0.3865.90 Safari/537.36"
7:53PM ERR authenticate: unexpected session state error="internal/sessions: session is not found" fwd_ip=["89.176.XX.XX"] ip=172.18.0.27 req_id=6045fcd9-c746-cc02-5ca2-a8349c8f4573 user_agent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/77.0.3865.90 Safari/537.36"
7:53PM DBG http-request duration=0.287996 email= fwd_ip=["89.176.XX.XX"] group= host=auth.example.com ip=172.18.0.27 method=GET path=/.pomerium/sign_in?redirect_uri=https%3A%2F%2Ffwdauth.example.com%2F.pomerium%2Fverify%2Fwhoami.example.com&sig=9_w8Odk7OgfbjO4RbnysOWTMJ-v09ocvIHNdOJII6q8%3D&ts=1570737223 req_id=6045fcd9-c746-cc02-5ca2-a8349c8f4573 service=all size=737 status=302 user_agent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/77.0.3865.90 Safari/537.36"
7:53PM DBG http-request duration=102.649998 email= fwd_ip=["89.176.XX.XX"] group= host=auth.example.com ip=172.18.0.27 method=GET path=/oauth2/callback?code=iBkxq2qABUTX1Sbl&state=U2VaM3dEZzZhTGxHem5zZkh5QjBHYTFmUzFiQkF4WkE5VXYwS1FQSWtBeXBjOGJQdUJ2THh2ZmlHeXpqNjFoODI4dXNqOGlhWHNDQkhlRWt0M2Ntb2c9PXwxNTcwNzM3MjIzfHcI1o_raY89tVDZ6PMj-dZ9gUrGd-fqIAN5kykXOVHFoncK8GQLyQCAmgy2YRdCgdNSYcxlHKzyGBvu7jZFBUg6vKM1TomYfqosMcHgs93nNe6FoaCbEDXH_4kcVIvQJwKP6Xgb_JMNxQK5_FwnGdOX_1BpfSoVJcrv7ypKGMd5KkXz7W4MbwxXU26PSDdo9R8UlKfxVpvGZonQnR3iEweBPdYpxxL75NuXKmpi7Wi-c8kqAHArG0ddP2M8tTm_00kO1vf9JbKd5YpPgbJ6CctnWoz1dlTAQQiTiXp1SIMSSJex1EmTdk9PQhuwdA_VxoDA req_id=f08ceb55-664f-61e7-e0c0-df0e405cdb63 service=all size=233 status=302 user_agent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/77.0.3865.90 Safari/537.36"
7:53PM DBG http-request duration=0.190173 email= fwd_ip=["89.176.XX.XX"] group= host=auth.example.com ip=172.18.0.27 method=GET path=/.pomerium/sign_in?redirect_uri=https%3A%2F%2Ffwdauth.example.com%2F.pomerium%2Fverify%2Fwhoami.example.com&sig=9_w8Odk7OgfbjO4RbnysOWTMJ-v09ocvIHNdOJII6q8%3D&ts=1570737223 req_id=88abd3d8-04c1-b654-9888-fb65c7b6f639 service=all size=118 status=302 user_agent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/77.0.3865.90 Safari/537.36"
7:53PM DBG http-request duration=0.63157 email=foo@example.com fwd_ip=["89.176.XX.XX"] group= host=fwdauth.example.com ip=172.18.0.27 method=GET path=/.pomerium/verify/whoami.example.com?pomerium-auth-callback=true req_id=30a20d0c-0149-2453-6a10-58f64708d9a5 service=all size=61 status=302 user_agent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/77.0.3865.90 Safari/537.36"
7:53PM DBG http-request duration=0.536717 email=foo@example.com fwd_ip=["89.176.XX.XX"] group= host=fwdauth.example.com ip=172.18.0.27 method=GET path=/.pomerium/verify/whoami.example.com req_id=7d8b5c6b-1407-be76-af90-9a39dd6a6ed5 service=all size=53 status=200 user_agent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/77.0.3865.90 Safari/537.36
```

### PR DESCRIPTION
**Pull request by desimone**
### Forward auth fix
These changes fix forward-auth support to work with the new per-route session changes made in `v0.5.0`. Previously, it was possible to simply set a shared-domain scope session cookie on the authenticate service's domain (`authenticate.x.y.z`) which, following authentication flow, would be passed to the forward authentication endpoint for validation. In the new model, the session must be set on the specific domain (sometimes referred to as [SameSite](https://www.owasp.org/index.php/SameSite) whose access is being delegated by the forward proxy and pomerium. These changes add the necessary callback flow to enable the new session management flow. 

### Entire URL is now signed, not just redirect URL

Previously, only _part_ of the URLs passed between services was signed (`redirect_uri`), which is largely a history of the project. With these changes the entire URL is signed/hmac'd plus includes explicit expiry and issue timestamps in a similar style as our JWT tokens. This is important because in many configurations, such as forward auth, the only way to pass information between services or following callbacks is using  query params (headers/cookies are not forwarded).  This ensures the authenticity of those query params and those payloads.

### Release candidate image
You can find the latest build on dockerhub. 

```
pomerium/pomerium:PR389vRC1
sha256:b7ea311f275cda7a7a4496f152a4257bb37703a3d0adb05385c68543edaa117f
```

#### Further reading
- https://pomerium-io.slack.com/archives/CK92MUAES/p1573832306087000
- https://github.com/containous/traefik/issues/2820
- https://github.com/thomseddon/traefik-forward-auth/issues/11
- https://community.containo.us/t/forwardauth-problem/2046

### Detailed
- internal/httputil: add global header constants
- scripts: update programmatic access script to use pomerium prepended redirect_uri to avoid conflicts
- internal/urlutil: abstract query params into their own package as constants
- internal/urlutil: add `SignedURL` struct enabling signing all parts of the URI
- http: remove a extraneous redundant `nosniff` header definitions in tests
- authenticate: update CSP headers for CSS changes. 

## Related issues
- Fixes #363 

**Checklist**:
- [x] add related issues
- [x] updated unit tests
- [ ] updated CHANGELOG.md
- [x] ready for review

### CODE REVIEW

File: proxy/handlers_test.go [hunk 1]
```diff
@@ -235,8 +235,8 @@ func TestProxy_VerifyWithMiddleware(t *testing.T) {
 		wantStatus   int
 		wantBody     string
 	}{
-		{"good", opts, nil, http.MethodGet, "", "/", "https://some.domain.example", &mock.Encoder{}, &sessions.MockSessionStore{Session: &sessions.State{Email: "user@test.example", Expiry: jwt.NewNumericDate(time.Now().Add(10 * time.Minute))}}, clients.MockAuthorize{AuthorizeResponse: true}, http.StatusOK, ""},
-		{"good verify only", opts, nil, http.MethodGet, "", "/verify", "https://some.domain.example", &mock.Encoder{}, &sessions.MockSessionStore{Session: &sessions.State{Email: "user@test.example", Expiry: jwt.NewNumericDate(time.Now().Add(10 * time.Minute))}}, clients.MockAuthorize{AuthorizeResponse: true}, http.StatusOK, ""},
+		{"good redirect", opts, nil, http.MethodGet, "", "/", "https://some.domain.example", &mock.Encoder{}, &sessions.MockSessionStore{Session: &sessions.State{Email: "user@test.example", Expiry: jwt.NewNumericDate(time.Now().Add(10 * time.Minute))}}, clients.MockAuthorize{AuthorizeResponse: true}, http.StatusFound, "<a href=\"https://some.domain.example\">Found</a>.\n\n"},
```
Comment (desimone) [inline proxy/handlers_test.go]: shame

File: proxy/middleware.go [hunk 2]
```diff
@@ -41,24 +52,15 @@ func (p *Proxy) AuthenticateSession(next http.Handler) http.Handler {
 
 // authenticate authenticates a user and sets an appropriate response type,
 // redirect to authenticate or error handler depending on if err on failure is set.
-func (p *Proxy) authenticate(errOnFailure bool, w http.ResponseWriter, r *http.Request) error {
+func (p *Proxy) authenticate(w http.ResponseWriter, r *http.Request) error {
 	s, err := sessions.FromContext(r.Context())
 	if err != nil {
 		p.sessionStore.ClearSession(w, r)
-
```
Comment (desimone) [inline proxy/middleware.go]: Now callers responsibility to handle result of authenticate.

File: authenticate/handlers.go [hunk 3]
```diff
@@ -24,10 +24,10 @@ import (
 // CSPHeaders are the content security headers added to the service's handlers
 // https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/script-src
 var CSPHeaders = map[string]string{
-	"Content-Security-Policy": "default-src 'none'; style-src 'self'" +
-		" 'sha256-z9MsgkMbQjRSLxzAfN55jB3a9pP0PQ4OHFH8b4iDP6s=' " +
-		" 'sha256-qnVkQSG7pWu17hBhIw0kCpfEB3XGvt0mNRa6+uM6OUU=' " +
-		" 'sha256-qOdRsNZhtR+htazbcy7guQl3Cn1cqOw1FcE4d3llae0='; " +
+	"Content-Security-Policy": "default-src 'none'; style-src " +
+		"'sha256-spMkVDoBBY86p0RC1fBYwdnGyMypJM8eG57+p3VASyk=' " +
```
Comment (desimone) [inline authenticate/handlers.go]: CSS is currently broken on authenticate because of this :(

File: internal/urlutil/signed.go [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+// 		the url is not nil.
+func NewSignedURL(key string, uri *url.URL) *SignedURL {
+	return &SignedURL{uri: *uri, key: key, timeNow: time.Now} // uri is copied
+}
+
+// Sign creates a shared-key HMAC signed URL.
+func (su *SignedURL) Sign() *url.URL {
+	now := su.timeNow()
+	issued := newNumericDate(now)
+	expiry := newNumericDate(now.Add(5 * time.Minute))
```
Comment (desimone) [inline internal/urlutil/signed.go]: 😉

File: internal/urlutil/signed.go [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+	params := su.uri.Query()
+	params.Set(QueryHmacIssued, fmt.Sprint(issued))
+	params.Set(QueryHmacExpiry, fmt.Sprint(expiry))
+	su.uri.RawQuery = params.Encode()
+	params.Set(QueryHmacSignature, hmacURL(su.key, su.uri.String(), issued, expiry))
```
Comment (desimone) [inline internal/urlutil/signed.go]: 👀 entire body is hmac'd instead of _just_ `redirect_uri`

File: internal/urlutil/signed.go (hunk 4 above)
Comment (travisgroth) [reply internal/urlutil/signed.go]: 😆

File: proxy/handlers.go [hunk 6]
```diff
@@ -136,86 +149,140 @@ func (p *Proxy) registerFwdAuthHandlers() http.Handler {
...
-		fmt.Fprintf(w, fmt.Sprintf("Access to %s is allowed.", uri.Host))
+		fmt.Fprintf(w, "Access to %s is allowed.", uri.Host)
 	})
 
 }
 
-// Callback takes a `redirect_uri` query param that has been hmac'd by the
-// authenticate service. Embedded in the `redirect_uri` are query-params
+// Callback takes a `pomerium_redirect_uri` query param that has been hmac'd by the
+// authenticate service. Embedded in the `pomerium_redirect_uri` are query-params
```
Comment (travisgroth) [inline proxy/handlers.go]: For clarity / future sanity, maybe just reference the new constant definitions

File: authenticate/handlers_test.go [hunk 7]
```diff
@@ -140,7 +140,7 @@ func TestAuthenticate_SignIn(t *testing.T) {
 			}
 			uri.RawQuery = queryString.Encode()
 
-			r := httptest.NewRequest(http.MethodGet, "/?redirect_uri="+uri.String(), nil)
+			r := httptest.NewRequest(http.MethodGet, "/?pomerium_redirect_uri="+uri.String(), nil)
```
Comment (travisgroth) [inline authenticate/handlers_test.go]: Should pull this value from the new constants

File: authenticate/handlers.go (hunk 3 above)
Comment (travisgroth) [reply authenticate/handlers.go]: Any way we can generate this automatically as part of the build?  This will get broken again.

Comment (desimone) [reply authenticate/handlers.go]: Let me pull this out into a new PR. I have an idea of how we can avoid inlining, and avoid this more or less all-together.

Comment (desimone) [reply authenticate/handlers.go]: Addressed in #392

File: proxy/handlers.go [hunk 8]
```diff
@@ -32,27 +31,34 @@ func (p *Proxy) registerDashboardHandlers(r *mux.Router) *mux.Router {
...
 
 	// Authenticate service callback handlers and middleware
+	// callback used to set route-scoped session and redirect back to destination
+	// only accept signed requests (hmac) from other trusted pomerium services
 	c := r.PathPrefix(dashboardURL + "/callback").Subrouter()
-	// only accept payloads that have come from a trusted service (hmac)
 	c.Use(middleware.ValidateSignature(p.SharedKey))
-	c.HandleFunc("/", p.Callback).Queries("redirect_uri", "{redirect_uri}").Methods(http.MethodGet)
 
+	c.Path("/").HandlerFunc(p.ProgrammaticCallback).Methods(http.MethodGet).
```
Comment (desimone) [inline proxy/handlers.go]: In a future PR, I'd like to split out programmatic login and callback into their own set of handlers and router similar to forward auth.

File: cmd/pomerium/main.go [hunk 9]
```diff
@@ -162,7 +162,7 @@ func newGlobalRouter(o *config.Options) *mux.Router {
 	if len(o.Headers) != 0 {
 		mux.Use(middleware.SetHeaders(o.Headers))
 	}
-	mux.Use(log.ForwardedAddrHandler("fwd_ip"))
+	mux.Use(log.HeadersHandler(httputil.HeadersXForwarded))
```
Comment (desimone) [inline cmd/pomerium/main.go]: e.g.
```json
{
  "level": "debug",
  "X-Forwarded-For": [
    "192.168.65.3"
  ],
  "X-Original-Method": [
    "GET"
  ],
  "X-Original-Url": [
    "https://dashboard-fwdauth.k8s.imac.bdd.io/api/v1/login/status"
  ],
  "X-Real-Ip": [
    "192.168.65.3"
  ],
  "X-Sent-From": [
    "nginx-ingress-controller"
  ],
  "ip": "192.168.1.1",
  "user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/78.0.3904.108 Safari/537.36",
  "referer": "https://dashboard-fwdauth.k8s.imac.bdd.io/",
  "req_id": "918990dd-e034-eac2-9891-2740ae3066b1",
  "duration": 0.821061,
  "size": 55,
  "status": 200,
  "email": "bobbydesimone@gmail.com",
  "group": "",
  "method": "GET",
  "service": "all",
  "host": "fwdauth.imac.bdd.io",
  "path": "/verify?uri=https://dashboard-fwdauth.k8s.imac.bdd.io/api/v1/login/status",
  "time": "2019-11-24T14:08:18-08:00",
  "message": "http-request"
}
```

File: internal/sessions/middleware.go [hunk 10]
```diff
@@ -41,8 +43,8 @@ func retrieveFromRequest(r *http.Request, sessions ...SessionLoader) (*State, er
 			return state, err
 		}
 		if state != nil {
-			err := state.Verify(r.Host)
-			return state, err // N.B.: state is _not nil_
+			err := state.Verify(urlutil.StripPort(r.Host))
```
[the remark this answers is not in the document]
Comment (desimone) [reply internal/sessions/middleware.go]: Which, I think is reasonable considering sessions themselves are not port specific. 
https://stackoverflow.com/questions/1612177/are-http-cookies-port-specific

File: authenticate/handlers.go [hunk 11]
```diff
@@ -100,71 +101,84 @@ func (a *Authenticate) refresh(w http.ResponseWriter, r *http.Request, s *sessio
...
+		callbackURL, err = urlutil.ParseAndValidateURL(callbackStr)
 		if err != nil {
-			httputil.ErrorResponse(w, r, httputil.Error("", http.StatusBadRequest, err))
+			httputil.ErrorResponse(w, r, httputil.Error(err.Error(), http.StatusBadRequest, err))
 			return
 		}
+		jwtAudience = append(jwtAudience, callbackURL.Hostname())
+	} else {
+		// otherwise, assume callback is the same host as redirect
+		callbackURL, _ = urlutil.DeepCopy(redirectURL)
```
Comment (travisgroth) [inline authenticate/handlers.go]: Is that an error we should be checking?

File: internal/sessions/middleware.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (travisgroth) [reply internal/sessions/middleware.go]: 🤔 TIL

File: proxy/middleware.go [hunk 12]
```diff
@@ -30,7 +30,18 @@ func (p *Proxy) AuthenticateSession(next http.Handler) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		ctx, span := trace.StartSpan(r.Context(), "proxy.AuthenticateSession")
 		defer span.End()
-		if err := p.authenticate(false, w, r.WithContext(ctx)); err != nil {
+		if err := p.authenticate(w, r.WithContext(ctx)); err != nil {
+			s, _ := sessions.FromContext(r.Context())
```
Comment (travisgroth) [inline proxy/middleware.go]: Error value thrown away - should it be checked?

File: proxy/handlers.go (hunk 8 above)
Comment (desimone) [reply proxy/handlers.go]: #400

File: proxy/middleware.go (hunk 12 above)
Comment (desimone) [reply proxy/middleware.go]: Though safe in this context, the flow is pretty janky. Updated to instantiated err and state at the same time.

File: authenticate/handlers.go (hunk 11 above)
Comment (desimone) [reply authenticate/handlers.go]: https://github.com/pomerium/pomerium/blob/aac36a8bdb0b1a0639a5fda204d6b544281ef496/authenticate/handlers.go#L110

I think safe in this context but I'll add a comment.

**Pull request merged**